desc/magnetic_fields.py

```diff
@@ -1094,3 +1130,644 @@ def odefun(rpz, s):
+        R_lmn=None,
+        Z_lmn=None,
+        modes_R=None,
+        modes_Z=None,
+        NFP=1,
+        sym="auto",
+        rho=1,
+        name="",
+        check_orientation=True,
+    ):
+        assert callable(potential), "Potential must be callable!"
+        assert callable(potential_dtheta), "Potential derivative must be callable!"
+        assert callable(potential_dzeta), "Potential derivative must be callable!"
+
+        self._potential = potential
+        self._potential_dtheta = potential_dtheta
+        self._potential_dzeta = potential_dzeta
+        self._surface_grid = surface_grid
+        self._params = params
+
+        assert (
+            self.surface_grid.NFP == NFP
+        ), "NFP of surface must match NFP of surface_grid"
+
+        super().__init__(
+            R_lmn, Z_lmn, modes_R, modes_Z, NFP, sym, rho, name, check_orientation
+        )
+
+    @property
+    def surface_grid(self):
+        """Grid of points to evaluate surface current at."""
+        return self._surface_grid
+
+    @surface_grid.setter
+    def surface_grid(self, new):
+        if new != self._surface_grid:
+            self._surface_grid = new
+
+    @property
+    def params(self):
+        """Dict of parameters to pass to potential function and its derivatives."""
+        return self._params
+
+    @params.setter
+    def params(self, new):
+        if new != self._params:
+            if len(new) != len(self._params):
+                warnings.warn(
+                    "Length of new params is different from length of current params! "
+                    "May cause errors unless potential function is also changed.",
+                    UserWarning,
+                )
+            self._params = new
+
+    @property
+    def potential(self):
+        """Potential function, signature (theta,zeta,**params) -> potential value."""
+        return self._potential
+
+    @potential.setter
+    def potential(self, new):
+        if new != self._potential:
+            assert callable(new), "Potential must be callable!"
+            self._potential = new
+
+    @property
+    def potential_dtheta(self):
+        """Phi poloidal deriv. function, signature (theta,zeta,**params) -> value."""
+        return self._potential_dtheta
+
+    @potential_dtheta.setter
+    def potential_dtheta(self, new):
+        if new != self._potential_dtheta:
+            assert callable(new), "Potential derivative must be callable!"
+            self._potential_dtheta = new
+
+    @property
+    def potential_dzeta(self):
+        """Phi toroidal deriv. function, signature (theta,zeta,**params) -> value."""
+        return self._potential_dzeta
+
+    @potential_dzeta.setter
+    def potential_dzeta(self, new):
+        if new != self._potential_dzeta:
+            assert callable(new), "Potential derivative must be callable!"
+            self._potential_dzeta = new
+
+    def compute_magnetic_field(self, coords, params=None, basis="rpz", grid=None):
```

Comment:
if we want to reduce duplicated code we could make this a private standalone function and then have both methods call that, similar to some of the `Equilibrium` methods